59 accumulation-to-rate

EddyCMWF opened the pull request:
### Description

This addition calculates the rate from accumulated variables using the ..temporal.accumulation_to_rate method. It can handle step types:
1. start_to_step
2. start_of_forecast
3. start_of_day

It can handle xarrays that use CF conventions, and whether the time-dimension represents the valid_time or a forecast_reference_time+forecast_period.

Also include is a temporal.deaccumulate method (equivalent to setting step="step_length" in accumulation_to_rate). This calculates the rate with units "per step", but the units attribute is not updated.

To do:
Example notebooks


### Contributor Declaration

By opening this pull request, I affirm the following:

* All authors agree to the [Contributor License Agreement](https://github.com/ecmwf/codex/blob/main/Legal/contributor_license_agreement.md).
* The code follows the project's coding standards.
* I have performed self-review and added comments where needed.
* I have added or updated tests to verify that my changes are effective and functional.
* I have run all existing tests and confirmed they pass.

EddyCMWF commented:
Hi @sandorkertesz  and @andreas-grafberger ,
If you want to use this branch for testing, it available on the test pypi here: https://test.pypi.org/project/earthkit-transforms/0.5.3.2.dev28/

Also, you can see the updated documentation, with notebook examples, on the github pages version of the docs that I use for release testing:
https://ecmwf.github.io/earthkit-transforms/notebooks/temporal/05-accumulation-to-rate-examples.html

andreas-grafberger commented on src/earthkit/transforms/temporal/_rates.py:
```diff
@@ -0,0 +1,362 @@
+    if rate_units == "step_length":
+        rate_scale_factor = 1.0
+        rate_units_str = ""  # Do not append anything to units attribute
+        rate_name_suffix = "_per_step"
```
What would you think about exposing the `rate_name_suffix` as an optional argument to the user?
Similar to how aggregation functions expose the `how_label` argument.

I think this might make this function a bit easier to use in data pipelines, where explicit control over the xarray dataset's output structure can be helpful.

andreas-grafberger commented on src/earthkit/transforms/temporal/_rates.py:
```diff
@@ -0,0 +1,362 @@
+
+                # If first step is midnight, we set to NaN as no prior data
+                if midnight_mask.data[0]:
+                    dataarray[0] = xp.nan
```
I could be missing something here but will this possibly update the original input dataarray inplace? And should we make sure to operate on a copy instead?

andreas-grafberger commented on src/earthkit/transforms/temporal/_rates.py:
```diff
@@ -0,0 +1,362 @@
+
+            output = diff_data / rate_scale_factor
+
+        case "start_of_day":
```
This could be overly cautious, but I think there are two cases where users might accidentally provide insufficient data without realizing:

1. `if period_dim_array is not None:` This assumes the forecast steps don't exceed one day (since we don't reset accumulations at midnight). What would you think about a docstring note or a simple check that max(step) < 24h?
2. If the step dimension doesn't contain an explicit `00:00` time (e.g. if it has gaps like `[23:00, 01:00, 02:00]`), the current `midnight_mask.shift()` logic won't detect `01:00` as the first step of a new day and reset the accumulation, leading to incorrect (potentially negative) rate values. Would a simple error check for this make sense?

andreas-grafberger reviewed (commented):
Hi @EddyCMWF! Thanks a lot for this feature, this will be really useful directly in one of our projects! :)

I left a few minor comments and would be curious to hear what you think.

sandorkertesz commented on src/earthkit/transforms/temporal/_rates.py:
```diff
@@ -0,0 +1,362 @@
+    step : float | int, optional
+        Interval between consecutive time steps (in hours, unless 'step_units' is specified). If not provided, the function
+        will infer the step from the first two time steps in the data.
+    step_units : str, optional
```
Could it be a timedelta? Then step_units is not needed.

sandorkertesz commented on src/earthkit/transforms/temporal/_rates.py:
```diff
@@ -0,0 +1,362 @@
+        - "start_of_day": accumulation restarts at the beginning of each day (00:00 UTC).
+        Default is "start_of_step".
+    from_first_step : bool, optional
+        Only used if `accumulation_type` is "start_of_forecast". If True, the first time step's rate is
```
How does it behave if I have these forecast steps: 24h, 30h, 36h?

EddyCMWF commented on src/earthkit/transforms/temporal/_rates.py:
```diff
@@ -0,0 +1,362 @@
+    """
+    if isinstance(dataarray, xr.Dataset):
+        data_vars = {
+            var_name: _accumulation_to_rate_dataarray(
```
Addresses in 3d294dd and 73cede9

EddyCMWF commented:
Hi @sandorkertesz  and @andreas-grafberger ,
Thanks for your reviews, I've addressed all the points raised, at least I think I did. Can you give it a check and tick if your satisfied and I'll get it merged.
Thanks,
Eddy

andreas-grafberger commented on src/earthkit/transforms/temporal/_rates.py:
```diff
@@ -0,0 +1,390 @@
+                    first_step_of_day_mask,
+                    dataarray.sel(_indexer) / rate_scale_factor,
+                    output,
+                )
```
Is the current `start_of_day` behavior intended when the first timestep is midnight and `from_first_step` is True?

Right now the code explicitly sets the first value to NaN if the first step is midnight, but the shifted `first_step_of_day_mask` contains a NaN at the first position. The `xr.where` treats that as truthy, overwriting the NaN. So the first value ends up as a computed rate instead of remaining unknown.

```python
import numpy as np
import xarray as xr
from earthkit.transforms import temporal

data = xr.DataArray(
    [5.0, 6.0, 8.0],
    dims=["valid_time"],
    coords={"valid_time": [
        np.datetime64("2024-01-01T00:00"),
        np.datetime64("2024-01-01T01:00"),
        np.datetime64("2024-01-01T02:00"),
    ]},
    name="tp",
    attrs={"units": "m"},
)

out = temporal.accumulation_to_rate(
    data,
    time_dim="valid_time",
    accumulation_type="start_of_day",
    from_first_step=True,
)

print(out.values)
```

Should the first step stay NaN in this case, or is the current overwrite the desired behavior? The docstring for `from_first_step` currently still states `Only used if [...] and the first step is not midnight.`
